### PR TITLE
LocalEventListener improve logging for missing serializability

### PR DESCRIPTION
**Pull request by msladek**
https://synjira.atlassian.net/browse/CELDEV-665

improves logging and prevents logfile flooding.

### CODE REVIEW

File: src/main/java/com/celements/common/observation/listener/LocalEventListener.java [hunk 1]
```diff
@@ -15,26 +20,65 @@
...
   }
 
+  boolean checkSerializability(Class<? extends Event> type, Object source, Object data) {
+    boolean serializable = true;
+    String msg = "unable to notify remote event [{0}]";
+    if (!(source instanceof Serializable)) {
+      msg += ", source [{1}] not serializable";
+      serializable = false;
+    }
+    if (!(data instanceof Serializable)) {
```
Comment (fpichler) [inline src/main/java/com/celements/common/observation/listener/LocalEventListener.java]: "instanceof Serializable" is not sufficient for checkSerializability. Please test by serializing the data and catching the exception.

**Review by fpichler — CHANGES_REQUESTED** [review]

File: src/main/java/com/celements/common/observation/listener/LocalEventListener.java [hunk 2]
```diff
@@ -15,26 +22,81 @@
...
-    if (event.getClass().isAnnotationPresent(Local.class)) {
+    Class<? extends Event> type = event.getClass();
+    if (type.isAnnotationPresent(Local.class)) {
       LOGGER.info("skipping local event '{}'", event.getClass());
     } else {
       try {
         super.onEvent(event, source, data);
       } catch (Exception exc) {
-        LOGGER.error("failed to notify RemoteObservationManager", exc);
+        if (containsNotSerializableException(exc)) {
```
Comment (msladek) [inline src/main/java/com/celements/common/observation/listener/LocalEventListener.java:45]: **Note**: `NotSerializableException` is a checked exception and not in the signature of `onEvent`, hence I cannot simply add a `catch(NotSerializableException nse)` block but have to analyse the given `Exception`

**Review by ebeutler — APPROVED** [review]

File: src/main/java/com/celements/common/observation/listener/LocalEventListener.java [hunk 3]
```diff
@@ -15,26 +22,81 @@
 @Component(LocalEventListener.NAME)
 public class LocalEventListener extends org.xwiki.observation.remote.internal.LocalEventListener {
 
-  private static final Logger LOGGER = LoggerFactory.getLogger(LocalEventListener.class);
+  private Logger LOGGER = LoggerFactory.getLogger(LocalEventListener.class);
 
   public static final String NAME = "observation.remote";
 
+  final Map<Class<? extends Event>, LogCounter> logCountMap = new ConcurrentHashMap<>();
```
Comment (fpichler) [inline src/main/java/com/celements/common/observation/listener/LocalEventListener.java:29]: Please make pivate?

File: src/main/java/com/celements/common/observation/listener/LocalEventListener.java [hunk 4]
```diff
@@ -15,26 +22,81 @@
...
+      LOGGER.warn(msg, exc);
+    } else if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug(format(msg, type.getSimpleName(), source, data), exc);
     }
+    logCountMap.get(type).increment();
+  }
+
+  static class LogCounter {
+
+    final long time = System.currentTimeMillis();
```
Comment (fpichler) [inline src/main/java/com/celements/common/observation/listener/LocalEventListener.java:81]: Please make the fields private

**Review by fpichler — CHANGES_REQUESTED** [review]

**Review by fpichler — APPROVED** [review]

**Comment by fpichler** [comment]
I understand now the need to access the properies from testing. Yet, it is a smell and we must improve it later on. Please add an Jira-Issue.

**Pull request merged**